nixos/modules/system/boot/luksroot.nix

```diff
@@ -1081,6 +1096,35 @@ in
     boot.initrd.preLVMCommands = mkIf (!config.boot.initrd.systemd.enable) (commonFunctions + preCommands + concatStrings (mapAttrsToList openCommand preLVM) + postCommands);
     boot.initrd.postDeviceCommands = mkIf (!config.boot.initrd.systemd.enable) (commonFunctions + preCommands + concatStrings (mapAttrsToList openCommand postLVM) + postCommands);
 
+    boot.initrd.systemd.services = let devicesWithClevis = filterAttrs (device: _: (hasAttr device clevis.devices)) luks.devices; in
+      mkIf (clevis.enable && systemd.enable) (
+        (mapAttrs'
+          (name: _: nameValuePair "cryptsetup-clevis-${name}" {
+            wantedBy = [ "systemd-cryptsetup@${utils.escapeSystemdPath name}.service" ];
+            before = [
+              "systemd-cryptsetup@${utils.escapeSystemdPath name}.service"
+              "initrd-switch-root.target"
+              "shutdown.target"
+            ];
+            wants = [ "systemd-udev-settle.service" ] ++ optional clevis.useTang "network-online.target";
```

Comment:
systemd-udev-settle is broken by design and should not be used, *ever*. See https://github.com/NixOS/nixpkgs/issues/73095#issuecomment-749464895.
I don't know what you're trying to depend on, but this is most likely not the correct way.

Reply:
We use that to wait for TPM2 to be available. Do you know of another synchronization point for that ?

Reply:
If you know the exact device path (say under /dev/ or /sys) in advance you can add a dependency on a device unit (see `systemd.device(5)`). Otherwise you could listen for the specific udev events using `udevadm monitor -s <subsystem>/<devicetype>` in you script. There's an example [here](https://github.com/NixOS/nixpkgs/blob/1be0f6fd4d473783332b8ce7fd00188b25640343/nixos/modules/services/networking/wpa_supplicant.nix#L146).

Reply:
It [seems](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=fdc915f7f71939ad5a3dda3389b8d2d7a7c5ee66) TPM2 devices get linked to /dev/tpmrm<n>, so you should probably add a `Wants=dev-tpmrm0.device` dependency.

I don't know if systemd TPMs devices by default, though. If not you'll also have to add a `TAG+=systemd` rule on those devices.

Reply:
Ah, look what was just commited to systemd: https://github.com/systemd/systemd/commit/4e1f0037b85d1b3c272e13862f44eb35844a18b1

Reply:
Yea, in the future, we should definitely change this to use `tpm2.target`. At the moment, I don't know if we can add dependencies on `dev-tpmrm0.device` directly, because that would cause boot to delay waiting for that device to timeout in the event that one doesn't exist.

Reply:
The tpm2 target does exactly this: it [adds a dependecy](https://github.com/systemd/systemd/commit/4e1f0037b85d1b3c272e13862f44eb35844a18b1#diff-3ad7dd8f19768deff635dea766969499867516e16d4c5d1f8337af468c7c7e90) on `dev-tpmrm0.device`, so it will time out without a TPM.

Reply:
@rnhmjoj But `tpm2.target` is not in the initial transaction by default. It only gets added by `systemd-tpm2-generator` *if* it detects that the TPM2 exists in `/dev` or according to the firmware (via efi variables). Units aren't supposed to have `Wants=tpm2.target`, only `After=tpm2.target` so that they're ordered properly *when a TPM exists*.

> Early boot programs that intend to access the TPM2 device should hence order themselves after this target unit, but not pull it in.

Reply:
@rnhmjoj I don't *think* a TPM is required necessarily? I think the `.jwe` determines if a TPM is required.